Optimize derived queries to avoid unnecessary JOINs for association ID access

academey opened the pull request:
## Summary

This PR optimizes derived query methods to avoid unnecessary JOINs when accessing association IDs. The implementation consolidates approaches from both this PR and PR #3922 into a unified solution with improved abstraction.

## Background

Starting from Hibernate 6.4.1, entity path traversal (e.g., `author.id`) generates JOINs even when accessing only the ID field. This affects query performance and generates suboptimal SQL for common patterns like `findByAuthorId`.

## Solution

### Unified Architecture
- **PathOptimizationStrategy**: Interface for optimizing property path traversal
- **JpaMetamodelContext**: Metamodel abstraction compatible with AOT compilation
- **DefaultPathOptimizationStrategy**: Implementation using `SingularAttribute.isId()` for reliable ID detection

### Implementation
- Modified QueryUtils to use unified optimization strategy
- Updated JpqlUtils to share the same optimization logic  
- Eliminated code duplication between Criteria API and JPQL generation
- Uses proven ID detection approach from PR #3922

## Examples

**Before optimization:**
```sql
-- findByAuthorId(1L) 
SELECT b FROM Book b JOIN b.author a WHERE a.id = 1
```

**After optimization:**  
```sql
-- findByAuthorId(1L)
SELECT b FROM Book b WHERE b.author.id = 1
```

## Features

- Handles @ManyToOne relationships
- Supports owning-side @OneToOne relationships
- Works with composite ID scenarios  
- Compatible with derived queries: `findByAuthorId`, `countByProfileId`, etc.
- AOT compilation support
- Graceful fallback when metamodel unavailable

## Testing

- Added comprehensive unit tests for optimization strategy
- Integration tests verify SQL generation improvements
- Tested with both Hibernate and EclipseLink providers
- All existing tests continue to pass

## Performance Impact

This optimization eliminates unnecessary JOINs in common query patterns:
- Simpler SQL execution plans
- Better utilization of foreign key indexes
- Consistent behavior across query generation methods
- No breaking changes to existing functionality

## Compatibility

- Works with Hibernate 6.4+ where the issue manifests
- Maintains compatibility with other JPA providers
- No changes required for existing application code
- Backward compatible with all Spring Data JPA versions

Fixes #3349

mp911de commented:
This is similar to #3922. Now with the fixes in Hibernate in place we can continue with both pull requests.

academey commented:
@mp911de I've completed the unified implementation as discussed. The key changes include

- Introduced `PathOptimizationStrategy` interface for abstraction
- Created `DefaultPathOptimizationStrategy` that incorporates the approach from PR #3922
- Added `JpaMetamodelContext` to minimize JPA API exposure and support AOT scenarios
- Integrated comprehensive test cases covering various optimization scenarios

The implementation now provides a clean abstraction layer that can be extended for different optimization strategies while maintaining compatibility with AOT compilation.

mp911de commented:
Thanks @academey and @bukajsytlos for your collaboration on optimizing queries. The tests from #3922 are much closer to what is actually happening and run as integration test validating repository usage.

I went ahead and took inspiration from both pull requests (and took #3922 as base) to refine expression creation and relationship Id detection. Given the scope of these changes, I would like to ship the change with the upcoming milestone first and ask you to give it a test. If those changes prove useful and do not break existing applications, we would consider backporting these into the `3.5.x` development line and ship them with the September service release.

Also, as general feedback:

1. Refrain from general `Hi @christophstrobl, @scordio, @quaff,` mentions. It rather leads to us to postpone working on tickets. We get tons of notifications. At-mentions are good for leaving breadcrumbs and references or asking (https://github.com/spring-projects/spring-data-jpa/pull/3970#issuecomment-3180699584 is proper usage)
2. Looking at the description of this PR, individual changes, the design and test cases in particular, I get a strong sense that a lot of these changes have been generated with AI without considering brevity, usefulness and amount of changes. I might be wrong. In any case, describing and repeating all changes from the PR in a description isn't helpful, takes a long time to read and understand and the net worth is negative. With that, the PR description resembles rather a marketing campaign instead of helping us to learn about your design and ideas. 
3. Unifying two things that work on slightly different abstraction levels is all but a trivial task, especially for folks that don't work on the code base on a regular basis. Thank you for your collaboration, all good things come eventually to an end and the important bit is how perspective changes through learning and reiterating on the actual problem.

That being said, I'm closing this PR as being superseded by #3922.

academey commented:
@mp911de Thanks for the feedback and for handling this optimization through #3922. I understand the decision

Also I appreciate the guidance on PR practices. I'll keep descriptions more concise and avoid mass mentions in future contributions. The complexity of unifying different abstraction levels was indeed a challenging aspect of this work. I nealry don't have experiences of making contributions to open source, so I was not good at that. I'm sorry about that. Looking forward to contributing to the project again with these learnings in mind. Thanks for maintaining such a great project
